### ISSUE TITLE
Channels of ATC codec are swizzled (BGR instead of RGB)

### ISSUE THREAD

**Issue by mems** [GPUOpen-Tools/Compressonator#19]
*TL;DR*: When Compressonator generate a texture using ATC codec (and DDS container, with CLI) the result texture channels are BGR instead of RGB

----

For the following image: <img height="20" width="20" src="https://cloud.githubusercontent.com/assets/729275/20038302/6baa5946-a431-11e6-8051-6b1fcebb3a36.png" alt="image4x4"> (4x4 image, generated with bash and ImageMagick: `printf '\xFF\x00\x00\x69\x00\xFF\x00\xFF\xFF\x00\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\xFF\x00\x00\xFF\x00\xFF\x00\xFF\xFF\x00\x00\xFF\xFF\x00\x00\xFF\xFF\x00\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF\x00\xFF' | convert -depth 8 -size 4x4+0 rgba:- image4x4.png`)

Compress it using: `CompressonatorCLI.exe -nomipmap -fd ATC_RGBA_Explicit image4x4.png image4x4.dds`

When I using this compressed texture with a device like LG Nexus 5X (Adreno 418) the colors are incorrect (B and R channels are inverted)

If you take the last 8 bytes (the RGB part of ATC block compression, skipping container header(s) and alpha part of the block), you get: <code><b>1F 00 E0 07</b> CC CF C0 FF</code>.
If I read the 4 first bytes by hand (match block color0 and color1) I get:
<pre>
color0 = <b>1F 00</b> (LE) = 0x001F = (skip first bit),0b00000,0b00000,0b11111 = 0,0,31 => rgb(0,0,255) = <b>blue</b> (incorrect)
color1 = <b>E0 07</b> (LE) = 0x07E0 = 0b00000,0b111111,0b00000 = 0,63,0 => rgb(0,255,0) = <b>green</b> (correct)
</pre>

If I use an other tool (like Adreno Texture Tool - QCompress output to DDS or KTX) the result will be <code><b>00 7C E0 07</b> CC CF C0 FF</code> (I can also check with DXT5 which formats look very similar). I've got:
<pre>
color0 = <b>00 7C</b> (LE), 0x7C00 = (skip first bit),0b11111,0b00000,0b00000 = 31,0,0 = rgb(255,0,0) = <b>red</b> (correct)
color1 = <b>E0 07</b> (LE), 0x07E0 = 0b00000,0b111111,0b00000 = 0,63,0 = rgb(0,255,0) = <b>green</b> (correct)
</pre>

Besides, the only information I found about ATC codec format is a [paper about converting DXT to ATC](http://www.guildsoftware.com/papers/2012.Converting.DXTC.to.ATC.pdf).

----

I think the cause is from [`RGBA8888_CHANNEL_…`](https://github.com/GPUOpen-Tools/Compressonator/blob/e8938d68d2755bc727c2a4b97d9c799eec0cb0da/Compressonator/Header/Codec/Buffer/CodecBuffer.h#L80-L82) used in [`CCodec_ATC::CompressRGBBlock`](https://github.com/GPUOpen-Tools/Compressonator/blob/493e9f10beed36a0d721ca33b57dfd1d84c5983a/Compressonator/Source/Codec/ATC/Codec_ATC.cpp#L59-L61). `rgbBlock` is read as (how I read it):
<pre>
struct colorData{
	<b>char B;//RGBA8888_CHANNEL_B=0</b>
	char R;//RGBA8888_CHANNEL_G=1
	<b>char G;//RGBA8888_CHANNEL_R=2</b>
	char A;//RGBA8888_CHANNEL_A=3
}
struct rgbBlock colorData[16];
</pre>

**Comment by NPCompress** [comment]
Thank you for such detailed information. Did you try the suggested change and it fixed the issue?
We will look into the code segment identified and run tests to determine the best way to correct for the next release. 

Note: there are several swizzle options for the data. Some are when the source data is read into RGBA buffer, another for the Codec type when the data is been processed in Compress() which swizzles the RGB to BGR for specific legacy codecs, and not for recent ones like ASTC, BC6H and BC7 etc.., there is even a user defined option to swizzle data.! 

So as you can see too many swizzle options may end up with some codecs been mishandled, if not properly managed.

**Comment by mems** [comment]
To fix the issue, I have to pre-swizzle R and B channels with ImageMagick
```
convert image.png -channel rgba -alpha on -set colorspace RGB -separate -swap 0,2 -combine -define png:color-type=6 image_bgra.png
CompressonatorCLI.exe -nomipmap -fd ATC_RGBA_Explicit image_bgra.png image.dds
```
(I don't use any swizzle option)

Note: I use Compressonator for DXT5+DDS and I don't have this issue: `CompressonatorCLI.exe -nomipmap -fd DXT5 image.png image.dds`

**Comment by NPCompress** [comment]
Did you use  V2.3 of CompressonatorCLI.exe or did you use the master-source and build your own binary?

**Comment by mems** [comment]
I use the V2.3.2953 36860e1 the only one binary available

**Comment by cpcdoy** [comment]
After looking very quickly at the code I guess it should just be a matter of calling one of the `SWIZZLE_*(i)` macros on the `srcBlock` before calling the `CCodec_ATC::CompressRGBBlock` function. Am I right ? Because changing the values of the `RGBA8888_CHANNEL_*` defines isn't a good idea as it's used in many other codecs as well.

**Comment by NPCompress** [comment]
cpcdoy: you are correct, this is used for buffer class to handle different codec buffer types. A command line option for swizzle is available in CLI that can be used along with the selected compression.  You can try "-doswizzle" 

CompressonatorCLI.exe -nomipmap -doswizzle -fd ATC_RGBA_Explicit image.png image.dds

**Comment by NPCompress** [comment]
fixed in v2.4 release

**Comment by umustbeloggedintododat** [comment]
Still not working for me. It's odd...
![1987_porsche_962c_ext etc_dds_ARGB_8888](https://user-images.githubusercontent.com/54049600/95642158-9e1b3780-0a74-11eb-85b3-e5ba5227dec3.png)

**Comment by umustbeloggedintododat** [comment]
I FOUND OUT THE PROBLEM! This texture compression is actually SUPPOSED TO BE BGR, so this patch RUINED ALL THE TEXTURE!!!!!!!
